src/prepare.js

```diff
@@ -116,21 +107,26 @@ function prepareElement(element, errorHandler, context) {
   }
 }
 
-function prepare(element, options = {}, context = {}) {
+async function prepare(element, options = {}, context = {}) {
   const {
     errorHandler = (error) => {
       throw error;
     },
   } = options;
-  return prepareElement(element, errorHandler, context).then(
-    ([children, childContext, p]) =>
-      Promise.all(
-        React.Children.toArray(children)
-          .map((child) => prepare(child, options, childContext))
-          .concat(p)
-          .filter(Boolean),
-      ),
```

Comment:
tbh, I like this alot more. It's a bit more readable and understandable in my eyes.
Refactoring out the `prepareElement(...).then(` is alot better, but the rest is just creating unnecessary variables and control flow imo.

Reply:
I thought the part where it uses `.concat(p)` with a `p` that is possibly undefined, and then `.filter(Boolean)` to remove the (possible) undefined was a bit confusing. What do you think about this instead? 
```javascript
  await Promise.all(
    React.Children.toArray(children)
      .map((child) => prepare(child, options, childContext))
      .concat(preparePromise ?? Promise.resolve()),
  );
```